vignettes/data-as-shiny-module.Rmd

```diff
@@ -0,0 +1,127 @@
+
+## Creating Data In-App
+
+One case for postponing data operations are data sets that are dynamic, frequently updated.
+Such data cannot be created once and kept in the global environment.
+Using `teal_data_module` allows to create a data set from scratch every time the user starts the application.
+This the user will always have access to the most recent version of the data.
+
+```{r, message = FALSE, warning = FALSE}
+library(teal)
+```
+
+
+```{r}
+data_mod <- teal_data_module(
+  ui = function(id) div(),
+  server = function(id) {
+    moduleServer(id, function(input, output, session) {
+      reactive({
+        data <- within(
+          teal.data::teal_data(),
+          {
+            dataset1 <- iris
+            dataset2 <- mtcars
+          }
+        )
+        teal.data::datanames(data) <- c("dataset1", "dataset2")
+        data
+      })
+    })
+  }
+)
+
+
+app <- init(
+  data = data_mod,
+  module = list(example_module())
+)
+
+if (interactive()) {
+  shinyApp(app$ui, app$server)
+}
+```
+
+
+## Modification Data In-App
+
+Another reason to postpone data operations is to allow the application user to act the preprocessing stage.
+An initial, constant form of the data can be created in the global environment and then modified once the app starts.
+
+The following example illustrates how `teal_data_module` can be utilized to subset data based on the user inputs:
+
+```{r}
+data <- within(teal_data(), {
+  dataset1 <- iris
+  dataset2 <- mtcars
+})
+teal.data::datanames(data) <- c("dataset1", "dataset2")
+
+data_mod <- teal_data_module(
+  ui = function(id) {
+    ns <- NS(id)
+    div(
+      selectInput(ns("species"), "Select species to filter",
+        choices = unique(iris$Species), multiple = TRUE
+      ),
+      actionButton(ns("submit"), "Submit")
+    )
+  },
+  server = function(id) {
+    moduleServer(id, function(input, output, session) {
+      eventReactive(input$submit, {
+        data_modified <- within(
+          data,
+          dataset1 <- subset(dataset1, Species %in% selected),
+          selected = input$species
+        )
+        data_modified
+      })
+    })
+  }
+)
+
+app <- init(
+  data = data_mod,
+  module = list(example_module())
+)
+
+if (interactive()) {
+  shinyApp(app$ui, app$server)
+}
+```
+
+_See `?within.qenv` for a detailed explanation of how to use the `within` method._
+
+Note that running preprocessing code in a module as opposed to the global environment will increase app loading times.
+It is recommended to keep the constant code in the global environment and to move only the dynamic parts to a data module.
+
+###### WARNING
+
+When using `teal_data_moduel` to modify a pre-existing `teal_data` object it is crucial that the server function and the data object are defined in the same environment as otherwise the server function will not be able to access the data object.
```

Comment:
```suggestion
When using `teal_data_module` to modify a pre-existing `teal_data` object it is crucial that the server function and the data object are defined in the same environment as otherwise the server function will not be able to access the data object.
```